Update `shared_priority_queue_scheduler`

biddisco opened the pull request:
This fixes segfaults in the shared_priority_scheduler and adds some general cleanup to affected code.

msimberg commented on libs/pika/schedulers/include/pika/schedulers/queue_holder_thread.hpp:
```diff
@@ -464,7 +468,7 @@ namespace pika { namespace threads { namespace policies {
         {
             PIKA_ASSERT(data.stacksize >= thread_stacksize::minimal);
-            PIKA_ASSERT(data.stacksize <= thread_stacksize::maximal);
+            PIKA_ASSERT(data.stacksize <= thread_stacksize::nostack);
```
Do you think it would make sense to reorderd the stacksizes so that `nostack == minimal`?

biddisco replied:
it probably would make sense, but that ought to be done in a separate PR, so I will leave this 'as is' for now

msimberg replied:
https://github.com/pika-org/pika/issues/213

msimberg commented on libs/pika/schedulers/include/pika/schedulers/shared_priority_queue_scheduler.hpp:
```diff
@@ -647,25 +639,28 @@ namespace pika { namespace threads { namespace policies {
         }
 
-        /// Return the next thread to be executed, return false if none available
+        // just a dummy function to keep the scheduling loop happy
```
This is not completely dummy, right? It just doesn't do the `wait` part? Though I'm not even sure the `wait` in `wait_or_add_new` makes sense for other schedulers either...

msimberg commented on libs/pika/schedulers/include/pika/schedulers/thread_queue_mc.hpp:
```diff
@@ -260,14 +246,17 @@ namespace pika { namespace threads { namespace policies {
         bool get_next_thread(threads::thread_id_ref_type& thrd, bool other_end,
             bool check_new = false) PIKA_HOT
         {
+            PIKA_MAYBE_UNUSED auto scp =
```
Same here about `[[maybe_unused]]`.

biddisco commented on libs/pika/debugging/include/pika/debugging/print.hpp:
```diff
@@ -319,8 +350,8 @@ namespace pika { namespace debug {
             ((tempstream << args << " "), ...);
-            tempstream << std::endl;
-            std::cout << tempstream.str();
+            tempstream << "\n";
```
the intention here is to force a `std::flush` after writing the contents to `cout` - to make it explicit, I have changed it to 
```
            tempstream << "\n";
            std::cout << tempstream.str() << std::flush();
```

biddisco commented:
I've cleaned up all the bits you marked except for the `nostack` one. Mostly these were stray bits left over from my larger changes that I removed prior to making the PR.

msimberg commented:
bors try

msimberg commented on libs/pika/debugging/include/pika/debugging/print.hpp:
```diff
@@ -19,6 +19,20 @@
 #include <type_traits>
 #include <utility>
 #include <vector>
+//
+#if defined(__linux) || defined(linux) || defined(__linux__)
+#include <sys/mman.h>
+#include <unistd.h>
+#elif defined(__APPLE__)
+#include <crt_externs.h>
+#include <unistd.h>
+#define environ (*_NSGetEnviron())
+#elif defined(PIKA_WINDOWS)
+#include <winsock.h>
+#define environ _environ
+#else
+extern char** environ;
+#endif
```
Sorry, one last thing. This looks unused and duplicates what's in `pika/debugging/environ.hpp`. Also a leftover?
```suggestion
```

msimberg replied:
Feel free to press "Commit suggestion" if this should be removed.

biddisco commented:
bors try

biddisco commented:
bors try
